src/format_rules.ml

```diff
@@ -0,0 +1,82 @@
+open Import
+
+let flag_of_kind : Ml_kind.t -> _ =
+  function
+  | Impl -> "--impl"
+  | Intf -> "--intf"
+
+let config_includes (config : Dune_file.Auto_format.t) s =
+  match config.enabled_for with
+  | Default -> true
+  | Only set -> List.mem s ~set
+
+let diff file1 file2 =
+  Action.Diff
+    { optional = false
+    ; mode = Text
+    ; file1
+    ; file2
+    }
+
+let add_diff sctx loc ~dir input output =
+  let module SC = Super_context in
+  let open Build.O in
+  let action = diff input output in
+  let alias_conf_name = "fmt" in
+  let stamp = ("format-alias", action) in
+  let alias = Build_system.Alias.make alias_conf_name ~dir in
+  SC.add_alias_action sctx alias ~loc:(Some loc) ~locks:[] ~stamp
+    (Build.paths [input; output]
+     >>>
+     Build.action
+     ~dir
+     ~targets:[]
+     action)
+
+let gen_rules sctx (config : Dune_file.Auto_format.t) ~dir =
+  let loc = config.loc in
+  let files =
+    File_tree.files_of
+      (Super_context.file_tree sctx)
+      (Path.drop_build_context_exn dir)
+  in
+  let resolve_program s = Super_context.resolve_program sctx ~loc:(Some loc) s
+  in
+  Path.Set.iter files ~f:(fun file ->
+    let input_basename = Path.basename file in
+    let target_s = input_basename ^ ".formatted" in
+    let input = Path.relative dir input_basename in
+    let output = Path.relative ~error_loc:loc dir target_s in
+    let context = Super_context.context sctx in
+
+    let ocaml kind =
+      if config_includes config Ocaml then
+        let exe = resolve_program "ocamlformat" in
```

Comment:
We could cache this outside of the `Path.Set.iter`, to avoid recomputing it for every file.